src/core/experimental/refcount.d

```diff
@@ -0,0 +1,545 @@
+
+        // { Get a const obj
+        @nogc nothrow pure @safe scope
+        this(return scope ref typeof(this) rhs) const
+        {
+            mixin(copyCtorIncRef);
+        }
+
+        @nogc nothrow pure @safe scope
+        this(return scope const ref typeof(this) rhs) const
+        {
+            mixin(copyCtorIncRef);
+        }
+
+        @nogc nothrow pure @safe scope
+        this(return scope immutable ref typeof(this) rhs) const
+        {
+            mixin(copyCtorIncRef);
+        }
+        // } Get a const obj
+
+        // { Get an immutable obj
+        @nogc nothrow pure @trusted scope
+        this(return scope ref typeof(this) rhs) immutable
+        {
+            // Can't have an immutable ref to a mutable. Create a new RC
+            rc = rhs.rc;
+            auto sz = rhs.payload.length;
+            int[] tmp = (cast(int*) pureAllocate(sz * int.sizeof))[0 .. sz];
+            tmp[] = rhs.payload[];
+            payload = cast(immutable) tmp;
+        }
+
+        @nogc nothrow pure @safe scope
+        this(return scope const ref typeof(this) rhs) immutable
+        {
+            rc = rhs.rc;
+            if (rhs.rc.isShared)
+            {
+                // By implementation, only immutable RC is shared, so it's ok to inc ref
+                payload = (() @trusted => cast(immutable) rhs.payload)();
+            }
+            else
+            {
+                // Can't have an immutable ref to a mutable. Create a new RC
+                auto sz = rhs.payload.length;
+                int[] tmp = (() @trusted => (cast(int*) pureAllocate(sz * int.sizeof))[0 .. sz])();
+                tmp[] = rhs.payload[];
+                payload = (() @trusted => cast(immutable) tmp)();
+            }
+        }
+
+        @nogc nothrow pure @safe scope
+        this(return scope immutable ref typeof(this) rhs) immutable
+        {
+            mixin(copyCtorIncRef);
+        }
+        // } Get an immutable obj
+
+        @nogc nothrow pure @safe scope
+        ref TestRC opAssign(return ref typeof(this) rhs)
+        {
+            if (payload is rhs.payload)
+            {
+                return rhs;
+                //return this;
+            }
+            if (rc.isInitialized && rc.isUnique)
+            {
+                () @trusted { pureDeallocate(payload); }();
+            }
+            payload = rhs.payload;
+            rc = rhs.rc;
+            return rhs;
+            //return this;
+        }
+
+        @nogc nothrow pure @trusted scope
+        ~this()
+        {
+            if (rc.isInitialized() && rc.isUnique())
+            {
+                pureDeallocate(cast(int[]) payload);
+            }
+        }
+    }
+
+    () @safe @nogc pure nothrow scope
+    {
+        enum numElem = 10;
+        auto t = TestRC(numElem);
+        assert(t.rc.isUnique);
+        const TestRC ct = const TestRC(numElem);
+        assert(ct.rc.isUnique);
+        immutable TestRC it = immutable TestRC(numElem);
```

Comment:
What happens when a return value of a pure function is implicitely converted to immutable?

Reply:
Can you please give me an example to help me understand the problem?

Reply:
Something like this
```
struct _RefCount { int* prc; }

struct S
{
	_RefCount rc;
	int payload;
}

S* createS(int x) pure
{
	S* s = new S;
	s.payload = x;
	return s;
}

void main()
{
	immutable(S)* ps = createS(3);
}
```

Reply:
@rainers if you want to use `__RefCount` you shouldn't use pointers as you'll bypass calls to the copy ctors/opAssign.

> Also works with values instead of pointers.

Changing your code to
> ```
> struct _RefCount { int* prc; }
> 
> struct S
> {
> 	_RefCount rc;
> 	int payload;
>      this(int x) { rc = _RefCount(1); payload = x; }
> }
> 
> S createS(int x) pure
> {
> 	S s = S(x);
> 	return s;
> }
> 
> void main()
> {
> 	immutable(S) ps = createS(3);
> }
> ```

I didn't know this will implicitly convert to immutable.
This could be problematic.

If this calls the copy ctor, then it should be fine; otherwise we'll have a problem.

Reply:
> If this calls the copy ctor, then it should be fine; otherwise I think we'll have a problem.

It should be subject to named value return optimisation (which means moving I think).

Reply:
This is a bummer. It means we need to always use interlocked inc/dec conservatively. I was hoping we'd do much better than C++'s `shared_ptr`.

Reply:
Pardon my ignorance, but why? It seems to me that moving a copy constructible should be no problem since the total number of instances stays the same and you only care about the refcount in a binary fashion, unless opPostMove messes with the refcount (why??) then all should be good.

Reply:
The issue is the implicit conversion to `immutable`.

The implementation allocates a buffer of two `uint`s: it uses the one aligned at 8 bytes for `immutable` and `shared` types, and the one aligned at 4 bytes for the rest. In order to check if we are dealing with a "possibly" `shared` reference we look at the alignment, and based on this we either do atomics or not.

In the pure factory function, we will create a non-shared `__RefCount`, so the internal counter will use the `uint` allocated at 4 bytes, thus marking the non-shared property. When this is returned and implicitly cast to `immutable`, the contract is broken.

The only solution to this, as far as we can tell, is to always assume `shared` and use atomics. This is what C++ had to do as well.